Security Annotations

martindrozdz opened the pull request:
- [X] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [X If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [X] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".


Resolves [#33118]

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java:
```diff
@@ -392,7 +396,9 @@ private void listTools(McpTransport transport) throws IOException {
         if (toolRegistry.hasTools()) {
             for (ToolMetadata tmd : toolRegistry.getAllTools()) {
-                response.add(new ToolDescription(tmd));
+                if (Authorizer.authorized(transport, tmd)) {
```
It looks to me like you're calling `Authorizer.authorise()` for callTool and listTools. We might not need it initialize and initialized notification, but I'm guessing we might also need it for cancelRequest

Azquelt replied:
The check for `cancelRequest` would be a bit different. Rather than checking that the user has the right role for the tool, we want to check that the request was created by the same user.

Currently we just rely on the Session ID, which should be secure since it's only revealed to the client which creates the session, but we may want to tie the session to the authenticated user and deny access to any request using that session ID but a different user.

Azquelt replied:
Since that falls outside role-based access, let's raise a separate issue for this.

ttmunyunguma commented on dev/io.openliberty.mcp.internal/bnd.bnd:
```diff
@@ -39,6 +39,7 @@ Bundle-Localization: \
 	com.ibm.websphere.org.osgi.service.component;version=latest,\
 	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
 	com.ibm.ws.logging;version=latest,\
+	io.openliberty.jakarta.annotation.2.1;version=latest,\
```
I wonder if you don't need annotations for -testpath as well?

Azquelt replied:
Everything on the build path is automatically available in tests as well, so that shouldn't be necessary.

ttmunyunguma commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/DenyAllTests.java:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    /*
+     * @Test
```
Did you mean to remove this?

martindrozdz replied:
Aah, I missed this. I can add another test class that will only list the tools for the whole package in the next PR.
I've removed this now.

ttmunyunguma reviewed (commented):
This is looking good. Great work covering all those multiple test scenarios

martindrozdz reviewed (commented):
Hi @ttmunyunguma ,

We are using:
- Authorizer.requireAuthorized(transport, tmd) for callTool and
- Authorizer.authorized(transport, tmd) for listTool

Good point about cancelRequest, I'm not sure how we will authorise those calls (I'll create a PR for it)

martindrozdz committed:
MCP - Security Annotations implementation

Applies authorization checks for MCP resources (specifically calling and
listing tools).

This is only tested on stateful mode only for this PR.

We look for the annotation DenyAll, PermitAll, RolesAllowed for all Tool
methods and classes.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AdminsRoleAllowedTests.java:
```diff
@@ -0,0 +1,198 @@
+import static com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions.SERVER_ONLY;
+import static org.junit.Assert.assertNotNull;
+
+import java.lang.System.Logger;
```
This is the wrong `Logger` class. Use `java.util.logging.Logger`.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AdminsRoleAllowedTests.java:
```diff
@@ -0,0 +1,198 @@
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "adminsRoleTools.war").addPackage(AdminsRoleTools.class.getPackage());
+        ShrinkHelper.exportDropinAppToServer(server, war, SERVER_ONLY);
+        server.startServer();
+        assertNotNull(server.waitForStringInLog("MCP server endpoint: .*/mcp$")); // regex matches string that ends with /mcp e.g. "MCP server endpoint: http://macbookpro.home:8010/toolTest/mcp"
```
I think this can be a `findStringsInLog` (faster in the case where the message is not found). The server shouldn't report as started until all the expected applications have started, and we should print that message before we start.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AdminsRoleAllowedTests.java:
```diff
@@ -0,0 +1,198 @@
+                          "CWMCM0014E", // An Internal Server Error occurred whilst processing the JSON-RPC request.
+                          "CWMCM0010E", //  Tool method threw an unexpected exception
+                          "CWMCM0011E" // An internal server error occurred
+        );
```
Do we actually expect all of these error messages in the logs for this test?

I imagine we would expect some of them, but others don't seem relevant to authentication.

martindrozdz replied:
I've kept these:
        server.stopServer(
                          "CWMCM0012E", // The requested JSON-RPC method is not found.
                          "CWMCM0014E", // An Internal Server Error occurred whilst processing the JSON-RPC request.
                          "CWMCM0011E" // An internal server error occurred
        );

Azquelt replied:
That's a better, shorter list, though I'm not sure why we're seeing those error messages for these tests. We should take a look at the logs.

Azquelt replied:
When I tested this locally, none of these messages appeared in the server log.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AdminsRoleAllowedTests.java:
```diff
@@ -0,0 +1,198 @@
+
+    @Test
+    public void testRolesAllowedClass_Admins_echoPermitAll() throws Exception {
+        String thisMethodName = Thread.currentThread().getStackTrace()[1].getMethodName();
```
You can use Junit's [TestName](https://junit.org/junit4/javadoc/4.12/org/junit/rules/TestName.html) to get the test method name.

martindrozdz replied:
I've used Testname and I get:
- "testDenyAllClass_EchoPermitAll_EE10_FEATURES" 
- instead of the actual test name which is testDenyAllClass_EchoPermitAll

This means the name of the test will be variable depending on the Jakarata EE version. This won't be reliable or future proof once the suffixes change in the future.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AuthHelper.java:
```diff
@@ -0,0 +1,115 @@
+        } catch (Exception e) {
+            String expectedErrorString = "Unexpected response: 403";
+            assertTrue("Forbidden 403 authorisation error expected", e.getMessage().contains(expectedErrorString));
+        }
```
Generally, we don't want to match against exception messages that we don't own like this because it makes the test brittle.

Someone else could change the wording of the exception message and then our test would start failing. (If we change our own messages, then we should update our own tests, but we shouldn't expect other teams to know that our tests depend on their messages not changing).

`HttpRequest` allows you to set the expected response code. Do that here instead.

martindrozdz replied:
I've added two different McpClient methods for this which now expect the 403 error:
- `callMCPAuthorisationErrorExpected(String jsonRequestBody)`
- `callMCPwithBasicAuth_AuthorisationErrorExpected(String jsonRequestBody, String user, String password)`

The response is a null in both cases so I assert that the response is null:
```
        String request = AuthHelper.buildMCPCallRequest();
        String response = client.callMCPAuthorisationErrorExpected(request);
        assertNull(response);
```

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AuthHelper.java:
```diff
@@ -0,0 +1,115 @@
+     */
+    static void assertAuthorisationErrorOnCall(McpClient client, String request, String user, String password) {
+        try {
+            client.callMCPwithBasicAuth(request, user, password);
+            fail("Forbidden 403 authorisation error expected");
+        } catch (Exception e) {
+            String expectedErrorString = "Unexpected response: 403";
+            assertTrue("Forbidden 403 authorisation error expected", e.getMessage().contains(expectedErrorString));
+        }
+    }
+
+    /*
+     * Given a JSON request and no authentication, then this MCP call is expected to fail authorization
+     */
+    static void assertAuthorisationErrorOnCall(McpClient client, String request) {
+        try {
+            client.callMCP(request);
+            fail("Forbidden 403 authorisation error expected");
+        } catch (Exception e) {
+            String expectedErrorString = "Unexpected response: 403";
+            assertTrue("Forbidden 403 authorisation error expected", e.getMessage().contains(expectedErrorString));
+        }
+    }
```
Should we expect 401 in this case?

martindrozdz replied:
I get a 403 returned which I check for.

Azquelt replied:
I can see that 403 is returned and that you check for it, I'm asking whether that's the _correct_ thing to be returned.

Usually, when someone needs to authenticate, you get a 401, whereas if you are authenticated but don't have access, you get a 403.

It's slightly odd here, the spec has some requirements around the response when you're not authorized, but they're all around using OAuth whereas this code is not specific to any authentication mechanism. We can discuss tomorrow.

Azquelt replied:
As per comment below, we raised a new issue for this.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/securityApps/AdminsRoleTools.java:
```diff
@@ -0,0 +1,85 @@
+ * Test app with @RolesAllowed at class level
+ * We expect that method level security annotations will override the class one
+ */
+public class AdminsRoleTools {
```
Could you move the annotations below the Javadoc please. I find them a bit easy to miss when they're above it like this.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/securityApps/AdminsRoleTools.java:
```diff
@@ -0,0 +1,85 @@
+        return input;
+    }
+
+    @RolesAllowed("Admins")
+    @Tool(name = "rolesAllowedClass_Admins_echoAdminAllowed", title = "Echoes the input", description = "Returns the input unchanged")
+    public String rolesAllowedClass_Admins_echoAdminAllowed(@ToolArg(name = "input", description = "input to echo") String input) {
+        return input;
+    }
```
In this class, there are several identical configurations, one for the allowed test and one for the forbidden test.

It may be clearer and more concise if we put all the scenarios which use one configuration into one test.

I think you're also missing a few scenarios. For example, `testRolesAllowedClass_Admins_echoDenyAll` tests that an unauthenticated user is denied access, but we don't test that a TestUser is denied access.

These missing scenarios would be more obvious if the tests for each configuration all followed the same pattern.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/securityApps/AdminsRoleTools.java:
```diff
@@ -0,0 +1,85 @@
+        return input;
+    }
+
+    @RolesAllowed("TestUsers")
+    @Tool(name = "rolesAllowedClass_Admins_echoTestUserAllowed", title = "Echoes the input", description = "Returns the input unchanged")
+    public String rolesAllowedClass_Admins_echoTestUserAllowed(@ToolArg(name = "input", description = "input to echo") String input) {
+        return input;
+    }
```
It's not a problem, but I'll note that there's no need to put all the metadata into each of these test tools.

You can get away with just
```java
    @RolesAllowed("TestUsers")
    @Tool
    public String rolesAllowedClass_Admins_echoTestUserAllowed(@ToolArg String input) {
        return input;
    }
```

Azquelt commented on dev/io.openliberty.mcp.internal_fat/publish/servers/mcp-server/server.xml:
```diff
@@ -4,8 +4,32 @@
+	     </group>
+	     
+	</basicRegistry>
+  
```
Can we do these tests with a separate server config please? (You just make a copy of the `mcp-server` directory, name it `mcp-server-auth` and update your test classes to have `@Server("mcp-server-auth")`)

This is important so that we test that `mcpServer-1.0` works when `appSecurity-5.0` is not enabled.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpTransport.java:
```diff
@@ -302,4 +302,23 @@ public <T> CompletionStage<Void> sendResultAsync(CompletionStage<T> stage) {
+        return req.isUserInRole(role);
+    }
+
+    /**
+     * Check if user is in role List
+     */
+    public boolean isUserInRole(List<String> roleList) {
```
```suggestion
    /**
     * Check if user has any of the roles in the List
     */
    public boolean isUserInRole(List<String> roleList) {
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorisationData.java:
```diff
@@ -0,0 +1,55 @@
+        METHOD_DENY, METHOD_PERMIT, METHOD_ROLES,
+        CLASS_DENY, CLASS_PERMIT, CLASS_ROLES,
+        NONE
+    }
```
I'm not sure we need to capture this level of detail. The distinction between the annotation being on the method or on the class doesn't change how we make the authorization decision, so it just means more cases to check.

There's also no distinction between `NONE` and `PERMIT`.

Additionally, when we implement programmatic registration of tools, this distinction won't make sense since the user will provide the permitted roles via a builder-style API.

martindrozdz replied:
This is where the authorization decision is made:

```
    private static boolean doAuthorization(McpTransport transport, ToolMetadata tmd) {
        AuthorisationData authData = tmd.authData();
        
        return switch (authData.getSecurityAnnotation()) {
            case NONE -> AUTHORISED;

            case METHOD_DENY -> NOT_AUTHORISED;
            case METHOD_PERMIT -> AUTHORISED;
            case METHOD_ROLES -> transport.isUserInRole(authData.getRoleList());

            case CLASS_DENY -> NOT_AUTHORISED;
            case CLASS_PERMIT -> AUTHORISED;
            case CLASS_ROLES -> transport.isUserInRole(authData.getRoleList());
        };
```

As you can see from the above, I need to discern between each annotation as the authorization decision is different for each one.

Azquelt replied:
> the authorization decision is different for each one.

The authorization decision is the same in many cases:
- for every `METHOD_XXX` case, the code to run is identical to the matching `CLASS_XXX` case.
- `NONE` is identical to `METHOD_PERMIT` and `CLASS_PERMIT`

There are only really three cases:
- `DENY`
- `PERMIT`
- `ROLES`

martindrozdz replied:
The main reason I have it this way to enforce the order of permission i.e. least permissive to most permissive.
Methods are tested first (Deny is least permissive, next permit then RolesAllowed) and then classes are tested in the same fashion. Method annotations override Class annotations so it is important to identify which is which.

Azquelt replied:
This isn't where you enforce the order of permission. You do that in `SecurityRequirement.createFrom`.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorisationData.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ *
+ */
+public class AuthorisationData {
```
Could we make this class a `record` please? I think it would simplify it and make it clearly immutable (which it should be).

Azquelt replied:
Please add javadoc describing the responsibilities of this class.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/DenyAllTests.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Test
+    public void testDenyAllClass_EchoAdminForbiddenWithoutAuthentication() throws Exception {
+        String thisMethodName = "testDenyAllClass_EchoAdminForbidden";
```
This is not the same method name.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorisationData.java:
```diff
@@ -0,0 +1,55 @@
+    public AuthorisationData(AnnotatedMethod<?> m) {
+        securityAnnotation = Authorizer.returnSecurityAnnotation(m);
+        roleList = getRoleList(m);
+    }
```
Why is the method for reading which security annotation in `Authorizer` but the method for reading the list of roles in this class?

martindrozdz replied:
Good catch. I've moved it to AuthorizationData and made it private.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AdminsRoleAllowedTests.java:
```diff
@@ -0,0 +1,198 @@
+    @Test
+    public void testRolesAllowedClass_Admins_echoNoSecurityAnnotationExists() throws Exception {
+        String thisMethodName = Thread.currentThread().getStackTrace()[1].getMethodName();
+        String request = AuthHelper.buildMCPCallRequest(thisMethodName);
+        AuthHelper.assertAuthorisationErrorOnCall(client, request, AuthHelper.GUEST, AuthHelper.VALID_PASSWORD);
+    }
+
+    @Test
+    public void testRolesAllowedClass_Admins_echoNoSecurityAnnotationExistsNoAuthentication() throws Exception {
+        String thisMethodName = "testRolesAllowedClass_Admins_echoNoSecurityAnnotationExists";
+        String request = AuthHelper.buildMCPCallRequest(thisMethodName);
+        AuthHelper.assertAuthorisationErrorOnCall(client, request);
+    }
```
You're not testing that this method permits access to an admin user.

martindrozdz replied:
This has now been remedied (See second AuthHelper.test):
```
    public void testRolesAllowedClass_Admins_echoNoSecurityAnnotationExists() throws Exception {
        AuthHelper.test(Scenario.NO_AUTHENTICATION, ExpectedTestResult.FAIL, client);
        AuthHelper.test(Scenario.ADMIN_PASS_LOGIN, ExpectedTestResult.PASS, client);
        AuthHelper.test(Scenario.ADMIN_FAIL_LOGIN, ExpectedTestResult.FAIL, client);
        AuthHelper.test(Scenario.TESTUSER_PASS_LOGIN, ExpectedTestResult.FAIL, client);
        AuthHelper.test(Scenario.TESTUSER_FAIL_LOGIN, ExpectedTestResult.FAIL, client);
        AuthHelper.test(Scenario.UNKNOWN_USER, ExpectedTestResult.FAIL, client);
        AuthHelper.test(Scenario.UNKNOWN_ROLE, ExpectedTestResult.FAIL, client);
}
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorisationData.java:
```diff
@@ -0,0 +1,55 @@
+            Collections.addAll(roleList, annotation.value());
+        }
+        return roleList;
+    }
```
This will return an empty list if the `RolesAllowed` annotation is on the class, rather than the method.

martindrozdz replied:
Excellent catch, I have updated AuthorizationData.createFrom to pass in the right annotation by checking the annotation type (Class or Method) before passing it into getRoleList(Annotated a).

    
```
    public enum AnnotationType {
        METHOD, CLASS, NONE
    }

    private static final Map<SecurityAnnotation, AnnotationType> ANNOTATION_TYPE_MAP = Map.of(SecurityAnnotation.METHOD_DENY, AnnotationType.METHOD,
                                                                                              SecurityAnnotation.METHOD_PERMIT, AnnotationType.METHOD,
                                                                                              SecurityAnnotation.METHOD_ROLES, AnnotationType.METHOD,
                                                                                              SecurityAnnotation.CLASS_DENY, AnnotationType.CLASS,
                                                                                              SecurityAnnotation.CLASS_PERMIT, AnnotationType.CLASS,
                                                                                              SecurityAnnotation.CLASS_ROLES, AnnotationType.CLASS,
                                                                                              SecurityAnnotation.NONE, AnnotationType.NONE);

public static AuthorizationData createFrom(AnnotatedMethod<?> m) {
        SecurityAnnotation securityAnnotation = collectSecurityAnnotation(m);
        AnnotationType annotationType = ANNOTATION_TYPE_MAP.get(securityAnnotation);

        List<String> computedRoleList = switch (annotationType) {
            //get roles list from class annotation (the method's declaring type)
            case CLASS -> {
                yield getRoleList(m.getDeclaringType());
            }
            //get roles list from method annotation
            case METHOD -> {
                yield getRoleList(m);
            }
            //return empty role list
            case NONE -> {
                yield new ArrayList<>();
            }
            //return empty role list
            default -> {
                yield new ArrayList<>();
            }
        };

        return new AuthorizationData(securityAnnotation, computedRoleList);
    }
```


Thank you.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,96 @@
+import jakarta.enterprise.inject.spi.AnnotatedType;
+import jakarta.servlet.http.HttpServletResponse;
+
+public class Authorizer {
```
Please add Javadoc explaining the responsibilities of this class.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,96 @@
+        } else {
+            throw new HttpResponseException(HttpServletResponse.SC_FORBIDDEN);
+        }
+    }
```
These methods are public and important, they should have Javadoc.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/securityApps/AdminsRoleTools.java:
```diff
@@ -0,0 +1,110 @@
+ * Test app with @RolesAllowed at class level
+ * We expect that method level security annotations will override the class one
+ */
+@ApplicationScoped
+@RolesAllowed("Admins")
+public class AdminsRoleTools {
```
Some of the tools in this class are no longer used.

Also, please order the tools in these classes in the same order as the tests that test them.

Azquelt replied:
As far as I can see, all the tools which have names ending `Forbidden` in this class are unused and should be removed.

Azquelt commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/securityApps/NoClassAnnotationTools.java:
```diff
@@ -0,0 +1,63 @@
+ * We expect that method level security annotations will override the class one
+ *
+ */
+@ApplicationScoped
+public class NoClassAnnotationTools {
```
This class is missing the case with no security annotation on the method.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorizationData.java:
```diff
@@ -0,0 +1,148 @@
+/**
+ * Used to store data relating to Authorization related to the Authorizer class in the same package
+ */
+public record AuthorizationData(SecurityAnnotation securityAnnotation, List<String> roleList) {
```
We should rename this class to something a little more specific. I recommend `SecurityRequirement`.

We also need to document what it _does_ rather than how it's used. (We still say how it's used, but we must document what it does, and that part should come first). I'd suggest
> Stores the requirements needed to access an MCP tool

We should also document the parameters here using `@param`.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorizationData.java:
```diff
@@ -0,0 +1,148 @@
+            Collections.addAll(roleList, annotation.value());
+        }
+        return roleList;
+    }
```
The `if` block here was added in case `a` is somehow `null` or doesn't have the expected annotation. I don't think that will ever happen with the current code but if it did, that would be an error and we should probably throw an exception.

However, this code to extract the role list could be moved into `collectSecurityAnnotation` which is doing the search for the annotations. Once you've found the annotation you can extract the role list, but since you already have the annotation, you don't need to check it exists first.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorizationData.java:
```diff
@@ -0,0 +1,148 @@
+
+    public List<String> getRoleList() {
+        return roleList;
+    }
```
You don't need these getter methods because this is a record so you automatically have getter methods, they're just named `securityAnnotation()` and `roleList()`.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,78 @@
+import io.openliberty.mcp.internal.exceptions.jsonrpc.HttpResponseException;
+import jakarta.servlet.http.HttpServletResponse;
+
+/**
+ * Used for user authorization for calling and listing of tools
+ */
+public class Authorizer {
```
```suggestion
/**
 * Performs checks for whether the user is authorized MCP resources.
 */
public class Authorizer {
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,78 @@
+    private static final boolean NOT_AUTHORIZED = false;
+
+    /**
+     *
+     * Authorizes the user without throwing an exception if authorization fails
+     *
+     * @param transport
+     * @param tmd
+     * @return true if the user has been authorized
+     */
+    public static boolean authorized(McpTransport transport, ToolMetadata tmd) {
+        return doAuthorization(transport, tmd);
+    }
```
- we should rename this to `isAuthorized` to make it more obvious that the user needs to check the return value
- we can make the Javadoc a bit more clear

```suggestion
    /**
     * Checks whether the user is authorized to access a particular MCP Tool.
     *
     * @param transport the McpTransport
     * @param tmd the tool to check access for
     * @return {@code true} if the user is authorized to access the tool
     */
    public static boolean isAuthorized(McpTransport transport, ToolMetadata tmd) {
        return doAuthorization(transport, tmd);
    }
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,78 @@
+
+    /**
+     *
+     * Authorizes the user and throws an exception if the user is not authorized
+     *
+     * @param transport
+     * @param tmd
+     * @return true if the user has been authorized
+     */
+    public static boolean requireAuthorized(McpTransport transport, ToolMetadata tmd) {
+        if (doAuthorization(transport, tmd)) {
+            return true;
+        } else {
+            throw new HttpResponseException(HttpServletResponse.SC_FORBIDDEN);
+        }
+    }
```
Since this method throws an exception if the condition it checks is not met, it should not return a `boolean`.

We can also improve the Javadoc

```suggestion
    /**
     * Throws an exception if the user is not authorized to access a particular MCP tool
     *
     * @param transport the McpTransport
     * @param tmd the tool to check access for
     * @throws HttpResponseException if the user is not authorized
     */
    public static void requireAuthorized(McpTransport transport, ToolMetadata tmd) {
        if (!doAuthorization(transport, tmd)) {
            throw new HttpResponseException(HttpServletResponse.SC_FORBIDDEN);
        }
    }
```

Azquelt replied:
Javadoc here is not correct

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java:
```diff
@@ -0,0 +1,78 @@
+     * -- @PermitAll
+     *
+     */
+    private static boolean doAuthorization(McpTransport transport, ToolMetadata tmd) {
```
We could move the content of this method into `isAuthorized`, but if we don't, we should add this paragraph to the start of the Javadoc:

```
     * Checks whether the user is authorized to access a particular MCP Tool.
```

Azquelt replied:
Hasn't been done.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java:
```diff
@@ -66,7 +67,8 @@ public record ToolMetadata(String name,
                            Function<ToolArguments, CompletionStage<ToolResponse>> asyncHandler,
-                           Optional<MethodMetadata> methodMetadata) {
+                           Optional<MethodMetadata> methodMetadata,
+                           AuthorizationData authData) {
```
Add `@param authData` to the Javadoc

Azquelt replied:
`@param authData` still missing from the Javadoc.

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/SecurityRequirement.java:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @param m the annotated method
+     * @return
+     */
```
This Javadoc is not attached to any method. Should it just be deleted?

martindrozdz replied:
I've removed the javadoc method you mentioned above

Re: requireAuthorized
You are looking at the wrong version of requireAuthorized:

 ```
   /**
     *
     * Throws an exception if the user is not authorized to access a particular MCP tool *
     *
     * @param transport
     * @param tmd
     * @return true if the user has been authorized
     */
    public static void requireAuthorized(McpTransport transport, ToolMetadata tmd) {
        if (!passedAuthorization(transport, tmd)) {
            throw new HttpResponseException(HttpServletResponse.SC_FORBIDDEN);
        }
    }
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/SecurityRequirement.java:
```diff
@@ -0,0 +1,107 @@
+import jakarta.enterprise.inject.spi.Annotated;
+import jakarta.enterprise.inject.spi.AnnotatedMethod;
+import jakarta.enterprise.inject.spi.AnnotatedType;
+
+/**
+ * Stores the authorization requirements data needed to authorize access to a MCP tool
+ *
+ * @param securityAnnotation the security annotation found that will be used for authorization decisions
+ * @param roleList the list of roles used for authorisation. This will be empty if no RolesAllowed annotation is being used to determine the authorization decision
+ */
+public record SecurityRequirement(SecurityAnnotation securityAnnotation, List<String> roleList) {
```
```suggestion
/**
 * Stores the authorization requirements data needed to authorize access to a MCP tool
 *
 * @param securityAnnotation the security annotation found that will be used for authorization decisions
 * @param roleList the list of roles allowed to access the tool. Having any of these roles is sufficient to grant access. This will be empty unless a RolesAllowed annotation is used to determine the authorization decision
 */
public record SecurityRequirement(SecurityAnnotation securityAnnotation, List<String> roleList) {
```

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/SecurityRequirement.java:
```diff
@@ -0,0 +1,107 @@
+            return new SecurityRequirement(SecurityAnnotation.METHOD_PERMIT, getRoleList(m));
+        if (m.isAnnotationPresent(RolesAllowed.class)) {
+            return new SecurityRequirement(SecurityAnnotation.METHOD_ROLES, getRoleList(m));
+        }
```
Why are we calling `getRoleList` when using `DenyAll` or `PermitAll` to create the security requirement?

It's also easier to see that the code is safe if we get hold of the `RolesAllowed` annotation within `createFrom`.


```suggestion
    /**
     * Return the roles allowed list for the annotation
     *
     * @param annotation the RolesAllowed annotation
     * @return the roleList for the given annotation
     */
    private static List<String> getRoleList(RolesAllowed annotation) {
        return Arrays.asList(annotation.value());
    }

    /**
     * Return the annotation type based on where the annotation resides.
     * This method enforces the order of permission i.e. least permissive to most permissive.
     * Methods are tested first (Deny is least permissive, next PermitAll then RolesAllowed)
     * Classes are tested in the same fashion.
     * Method annotations override Class annotations.
     *
     * @param m the annotated method
     * @return the Authorization data calculated
     */
    public static SecurityRequirement createFrom(AnnotatedMethod<?> m) {

        // Method-level checks (higher precedence)
        if (m.isAnnotationPresent(DenyAll.class))
            return new SecurityRequirement(SecurityAnnotation.METHOD_DENY, Collections.emptyList());
        if (m.isAnnotationPresent(PermitAll.class))
            return new SecurityRequirement(SecurityAnnotation.METHOD_PERMIT, Collections.emptyList());
        if (m.isAnnotationPresent(RolesAllowed.class)) {
            return new SecurityRequirement(SecurityAnnotation.METHOD_ROLES, getRoleList(m.getAnnotation(RolesAllowed.class)));
        }
```

martindrozdz replied:
Done

martindrozdz commented:
That version of Authorizer.doAuthorization() is outdated (You are looking at the wrong version).
The new one is Authorizer.passedAuthorization()
https://github.com/martindrozdz/open-liberty/blob/feature-33118-SecAnnos/dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java

martindrozdz commented:
I don't have assertAuthorisationErrorOnCall as a method anymore in AuthHelper. You are looking at an incorrect version of it.
(https://github.com/martindrozdz/open-liberty/blob/feature-33118-SecAnnos/dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AuthHelper.java)

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java:
```diff
@@ -55,6 +56,7 @@
  * @param handler the handler, {@code null} if this is an async tool
  * @param asyncHandler the async handler, {@code null} if this is not an async tool
  * @param methodMetadata data about the method if this tool was discovered by annotating a method with {@link Tool}. Should not be used in most circumstances.
+ * @param SecurityRequirement stores authorization requirements needed to authorize access to a MCP tool
```
```suggestion
 * @param authData stores authorization requirements needed to authorize access to a MCP tool
```

Azquelt commented:
> That version of Authorizer.doAuthorization() is outdated (You are looking at the wrong version). The new one is Authorizer.passedAuthorization() https://github.com/martindrozdz/open-liberty/blob/feature-33118-SecAnnos/dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/Authorizer.java

The new version still does not have the requested line at the start of the Javadoc:

```java
     * Checks whether the user is authorized to access a particular MCP Tool.
```

Edit: this is now done

Azquelt commented:
> I don't have assertAuthorisationErrorOnCall as a method anymore in AuthHelper. You are looking at an incorrect version of it. (https://github.com/martindrozdz/open-liberty/blob/feature-33118-SecAnnos/dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AuthHelper.java)

I am not looking at the incorrect version of it. The question still applies to the current version of the code. You've moved it [here](https://github.com/OpenLiberty/open-liberty/pull/33516/files#diff-546a1a9f8b538eaa135f4b1e74fbad3c59bc5af6db8693568d6d5d1ba262e721R192), but it still checks for a 403 status code when the user hasn't provided credentials, and I'm still not convinced that's correct.

Azquelt commented:
> > I don't have assertAuthorisationErrorOnCall as a method anymore in AuthHelper. You are looking at an incorrect version of it. (https://github.com/martindrozdz/open-liberty/blob/feature-33118-SecAnnos/dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/security/AuthHelper.java)
> 
> I am not looking at the incorrect version of it. The question still applies to the current version of the code. You've moved it [here](https://github.com/OpenLiberty/open-liberty/pull/33516/files#diff-546a1a9f8b538eaa135f4b1e74fbad3c59bc5af6db8693568d6d5d1ba262e721R192), but it still checks for a 403 status code when the user hasn't provided credentials, and I'm still not convinced that's correct.

We discussed this and agreed we probably should have a 401 if the user has not successfully authenticated and needs to do so. However, we're not immediately sure how to do this, so we'll raise a new issue for it.

Azquelt reviewed (approved):
This looks good, please squash the commits
